File: stats.py
"""Слой 4: обратная связь и статистика.

Недельный отчёт, месячный отчёт, стрик. Работает поверх history.
"""

from datetime import datetime, timedelta
from zoneinfo import ZoneInfo

import config
from db import from_iso, to_iso


def _local_date(iso_sent_at: str, tz: str):
    dt = from_iso(iso_sent_at)
    if dt is None:
        return None
    return dt.astimezone(ZoneInfo(tz)).date()


def compute_streak(store, user_id: int, tz: str) -> int:
    """Стрик: сколько дней подряд (считая от сегодня/вчера назад) было хотя бы
    одно выполненное задание ('done'). Пропущенный день обрывает стрик.
    Сегодня без активности стрик не обрывает, если вчера была — даём день добить.
    """
    rows = store.all_history(user_id)
    done_days = set()
    for r in rows:
        if r["status"] == "done":
            d = _local_date(r["sent_at"], tz)
            if d:
                done_days.add(d)
    if not done_days:
        return 0

    today = datetime.now(ZoneInfo(tz)).date()
    # Стартуем с сегодня, если есть done; иначе со вчера (день ещё не кончился).
    if today in done_days:
        cursor = today
    elif (today - timedelta(days=1)) in done_days:
        cursor = today - timedelta(days=1)
    else:
        return 0

    streak = 0
    while cursor in done_days:
        streak += 1
        cursor -= timedelta(days=1)
    return streak


def _count(rows, track_id, status):
    return sum(1 for r in rows if r["track_id"] == track_id and r["status"] == status)


def _sent(rows, track_id):
    # Сколько раз трек реально просили (любой финальный или отправленный статус).
    return sum(1 for r in rows if r["track_id"] == track_id)


def weekly_report(store, user_id: int, tz: str) -> str:
    now = datetime.now(ZoneInfo(tz))
    start_local = (now - timedelta(days=7)).replace(hour=0, minute=0, second=0,
                                                     microsecond=0)
    rows = store.history_between(
        user_id, to_iso(start_local.astimezone(ZoneInfo("UTC"))),
        to_iso(now.astimezone(ZoneInfo("UTC"))),
    )
    streak = compute_streak(store, user_id, tz)

    lines = ["📊 *Недельный отчёт*", ""]
    for tid in config.TRACK_ORDER:
        track = store.get_track(user_id, tid)
        if not track or not track["enabled"]:
            continue
        done = _count(rows, tid, "done")
        planned = _sent(rows, tid)
        title = config.TRACK_TITLES[tid]
        if tid == config.TRACK_EGE:
            mark = "✅ сделал" if done else "— не сделал"
            lines.append(f"• *{title}*: {mark}")
        else:
            lines.append(f"• *{title}*: сделано {done} из {planned or '—'}")

    lines.append("")
    fire = "🔥" if streak > 0 else "•"
    lines.append(f"{fire} Стрик: *{streak}* дн. подряд")
    if streak == 0:
        lines.append("_Один день — и стрик снова живой. Начни сегодня._")
    return "\n".join(lines)


def monthly_report(store, user_id: int, tz: str):
    """Возвращает (текст, предложение_адаптации|None).

    Предложение — (track_id, новый_n_days) если трек проседает.
    """
    now = datetime.now(ZoneInfo(tz))
    start_local = (now - timedelta(days=30)).replace(hour=0, minute=0, second=0,
                                                     microsecond=0)
    rows = store.history_between(
        user_id, to_iso(start_local.astimezone(ZoneInfo("UTC"))),
        to_iso(now.astimezone(ZoneInfo("UTC"))),
    )

    lines = ["🗓 *Месячный отчёт*", ""]
    total_done = total_planned = 0
    worst = None  # (ratio, track_id)
    for tid in config.TRACK_ORDER:
        track = store.get_track(user_id, tid)
        if not track or not track["enabled"]:
            continue
        done = _count(rows, tid, "done")
        planned = _sent(rows, tid)
        total_done += done
        total_planned += planned
        ratio = (done / planned) if planned else None
        title = config.TRACK_TITLES[tid]
        pct = f"{round(ratio * 100)}%" if ratio is not None else "нет данных"
        lines.append(f"• *{title}*: {done}/{planned or '—'} ({pct})")
        if ratio is not None and planned >= 3:
            if worst is None or ratio < worst[0]:
                worst = (ratio, tid)

    overall = (total_done / total_planned) if total_planned else 0
    lines.insert(2, f"Общий процент выполнения: *{round(overall * 100)}%*")
    lines.insert(3, "")

    suggestion = None
    if worst and worst[0] < config.UNDERPERFORM_RATIO:
        ratio, tid = worst
        track = store.get_track(user_id, tid)
        title = config.TRACK_TITLES[tid]
        lines.append("")
        if track["cadence"] == "every_n_days":
            new_n = track["n_days"] + 2
            lines.append(
                f"⚠️ *{title}* проседает ({round(ratio*100)}%). Может, реже —"
                f" раз в {new_n} дня вместо {track['n_days']}? Так проще держать ритм."
            )
            suggestion = (tid, new_n)
        else:
            lines.append(
                f"⚠️ *{title}* проседает ({round(ratio*100)}%). Стоит вернуться"
                " к нему или пересмотреть формат."
            )
    return "\n".join(lines), suggestion

File: worker/src/stats.js
// Слой 4: статистика — стрик, недельный и месячный отчёты.

import {
  TRACK_ORDER, TRACK_TITLES, TRACK_EGE, UNDERPERFORM_RATIO,
} from "./config.js";
import { localInfo } from "./scheduling.js";
import * as db from "./db.js";

function localDate(ms, tz) {
  return localInfo(ms, tz).dateStr;
}

// Прибавить дни к строке даты YYYY-MM-DD.
function addDaysStr(dateStr, days) {
  const [y, m, d] = dateStr.split("-").map(Number);
  const nd = new Date(Date.UTC(y, m - 1, d) + days * 86400000);
  const p = (n) => String(n).padStart(2, "0");
  return `${nd.getUTCFullYear()}-${p(nd.getUTCMonth() + 1)}-${p(nd.getUTCDate())}`;
}

export async function computeStreak(env, userId, tz) {
  const rows = await db.allHistory(env, userId);
  const doneDays = new Set();
  for (const r of rows) if (r.status === "done") doneDays.add(localDate(r.sent_at, tz));
  if (!doneDays.size) return 0;

  const today = localInfo(Date.now(), tz).dateStr;
  let cursor;
  if (doneDays.has(today)) cursor = today;
  else if (doneDays.has(addDaysStr(today, -1))) cursor = addDaysStr(today, -1);
  else return 0;

  let streak = 0;
  while (doneDays.has(cursor)) {
    streak += 1;
    cursor = addDaysStr(cursor, -1);
  }
  return streak;
}

function count(rows, trackId, status) {
  return rows.filter((r) => r.track_id === trackId && r.status === status).length;
}
function sent(rows, trackId) {
  return rows.filter((r) => r.track_id === trackId).length;
}

export async function weeklyReport(env, userId, tz) {
  const now = Date.now();
  const rows = await db.historyBetween(env, userId, now - 7 * 86400000, now);
  const streak = await computeStreak(env, userId, tz);

  const lines = ["📊 *Недельный отчёт*", ""];
  for (const tid of TRACK_ORDER) {
    const track = await db.getTrack(env, userId, tid);
    if (!track || !track.enabled) continue;
    const done = count(rows, tid, "done");
    const planned = sent(rows, tid);
    const title = TRACK_TITLES[tid];
    if (tid === TRACK_EGE) {
      lines.push(`• *${title}*: ${done ? "✅ сделал" : "— не сделал"}`);
    } else {
      lines.push(`• *${title}*: сделано ${done} из ${planned || "—"}`);
    }
  }
  lines.push("");
  lines.push(`${streak ? "🔥" : "•"} Стрик: *${streak}* дн. подряд`);
  if (!streak) lines.push("_Один день — и стрик снова живой. Начни сегодня._");
  return lines.join("\n");
}

export async function monthlyReport(env, userId, tz) {
  const now = Date.now();
  const rows = await db.historyBetween(env, userId, now - 30 * 86400000, now);

  const body = [];
  let totalDone = 0, totalPlanned = 0;
  let worst = null; // {ratio, tid}
  for (const tid of TRACK_ORDER) {
    const track = await db.getTrack(env, userId, tid);
    if (!track || !track.enabled) continue;
    const done = count(rows, tid, "done");
    const planned = sent(rows, tid);
    totalDone += done;
    totalPlanned += planned;
    const ratio = planned ? done / planned : null;
    const pct = ratio === null ? "нет данных" : `${Math.round(ratio * 100)}%`;
    body.push(`• *${TRACK_TITLES[tid]}*: ${done}/${planned || "—"} (${pct})`);
    if (ratio !== null && planned >= 3 && (!worst || ratio < worst.ratio)) {
      worst = { ratio, tid };
    }
  }
  const overall = totalPlanned ? totalDone / totalPlanned : 0;

  const lines = ["🗓 *Месячный отчёт*", "",
    `Общий процент выполнения: *${Math.round(overall * 100)}%*`, "", ...body];

  let suggestion = null;
  if (worst && worst.ratio < UNDERPERFORM_RATIO) {
    const track = await db.getTrack(env, userId, worst.tid);
    const title = TRACK_TITLES[worst.tid];
    lines.push("");
    if (track.cadence === "every_n_days") {
      const newN = track.n_days + 2;
      lines.push(`⚠️ *${title}* проседает (${Math.round(worst.ratio * 100)}%). Может, реже — раз в ${newN} дня вместо ${track.n_days}? Так проще держать ритм.`);
      suggestion = { tid: worst.tid, newN };
    } else {
      lines.push(`⚠️ *${title}* проседает (${Math.round(worst.ratio * 100)}%). Стоит вернуться к нему или пересмотреть формат.`);
    }
  }
  return { text: lines.join("\n"), suggestion };
}
